### ISSUE TITLE
Feature request - keycast - show keybindings and commands in modeline

### ISSUE THREAD

**Issue by practicalli-john** [syl20bnr/spacemacs#14169]
The https://github.com/tarsius/keycast project displays the Emacs key binding and its command in the mode line.  This is very useful for demos and screencasts to show off the features of Spacemacs, without taking up space in the Emacs frame.

![keycast example](https://camo.githubusercontent.com/08fa0ec66da70f182459f8a68baf79584a9de9d5253cdfb73c75d5fda543ab84/687474703a2f2f726561646d652e656d6163736169722e6d652f6b6579636173742e706e67)

command-log (already part of Spacemacs) does show the history of key bindings and commands, although this does take up a significant part of the frame, even if the font is made small.

Should this package be added to Spacemacs directly, or would it be better as its own layer?

**A working configuration - hack**
I have keycast working using the doom-modeline theme and Spacemacs develop.
I added keycast as an `dotspacemacs-additional-package` and added the following code in `dotspacemacs/user-config` which I assume could be greatly improved.
```
  (use-package keycast
                :commands keycast-mode
                :config
                (define-minor-mode keycast-mode
                  "Show current command and its key binding in the mode line."
                  :global t
                  (if keycast-mode
                      (progn
                        (add-hook 'pre-command-hook 'keycast-mode-line-update t)
                        (add-to-list 'global-mode-string '("" mode-line-keycast " "))
                        )
                    (remove-hook 'pre-command-hook 'keycast-mode-line-update)
                    (setq global-mode-string (remove '("" mode-line-keycast " ") global-mode-string)))))
```

I took this code from https://tecosaur.github.io/emacs-config/config.html#fun,code--8 although had to tweak a few commands and remove the custom faces from that code as it interfered with the flycheck tool-tips, which I use to see the linter output in the source code (using clj-kondo).

**Comment by robbyoconnor** [comment]
@jr0cket, Why make the layer if you have the time do so, if not that is okay.

**Comment by smile13241324** [comment]
I think this would be pretty useful however to keep layer clutter at a minimum I would prefer to have this integrated into `command-log-layer` which should all be about packages which allow proper demonstrations with Spacemacs.

**Comment by jr0cket** [comment]
I can see the value of keycast being part of a command log layer.  I was using command-log before keycast to share my keybindings

I switched to keycast as it uses less space than command-log and the colour used for command-log did not work with light themes

I am afraid I am unlikely to find time in the next couple of months to try work on this though

**Comment by robbyoconnor** [comment]
@jr0cket sounds good to me!

**Comment by smile13241324** [comment]
I just had a look into this, looks like we need to invest some time into this to work properly with the different supported mode lines. For now it is not working in spaceline for example.

I have added the following config in the layer, but would require a proper value for `keycast-insert-after` to work as it should. Also when I was working with split windows only one window had a working mode line after activation key cast.

Here is the config I have stitched together which works rudimentary with spaceline. For other modes we would require similar or better strings. Also the mode line should also work in a multi window scenario.

If someone of you find the time to tackle this it would be very much appreciated. If not I may commit at least this basic integration and let others add support for the individual mode lines.

``` elisp
;;; packages.el --- command-log Layer packages File for Spacemacs
;;
;; Copyright (c) 2012-2021 Sylvain Benner & Contributors
;;
;; Author: Sylvain Benner <sylvain.benner@gmail.com>
;; URL: https://github.com/syl20bnr/spacemacs
;;
;; This file is not part of GNU Emacs.
;;
;;; License: GPLv3

(defconst command-log-packages
  ;; bmag/command-log-mode is a fork of lewang/command-log-mode, and there
  ;; is an open PR to merge the fork into the original repo.
  ;; TODO when the PR is merged upstream, change to use the original package
  ;; from MELPA (IOW remove :location argument)
  '((command-log-mode :location (recipe :fetcher github
                                        :repo "bmag/command-log-mode"
                                        :branch "color"))
    keycast))

(defun command-log/init-command-log-mode ()
  (use-package command-log-mode
    :commands global-command-log-mode
    ;; :commands (clm/open-command-log-buffer global-command-log-mode spacemacs/toggle-command-log-mode)
    :init
    (spacemacs/declare-prefix "atl" "command log")
    (spacemacs/set-leader-keys "atll" #'global-command-log-mode)
    :config
    (setq clm/log-command-exceptions* (append clm/log-command-exceptions*
                                              '(evil-next-line
                                                evil-previous-line
                                                evil-forward-char
                                                evil-backward-char))
          command-log-mode-auto-show t)))

(defun command-log/init-keycast()
  (use-package keycast
    :init (spacemacs/set-leader-keys "atlk" #'keycast-mode)
    :config
    (setq keycast-insert-after "%e")))
```

**Comment by practicalli-johnny** [comment]
I noticed keycast has been added to the command-log layer, thank you.

Initial support for keycast added in spacemacs via https://github.com/syl20bnr/spacemacs/commit/bf32e4333916e3790f9a1ab225a896f03bf9a6b7

I use doom modeline and enabling keycast-mode replaces the whole modeline with keycast (although otherwise seems to work very nicely).

Update: Having tried keycast with the default spacemacs modeline, keycast also replaces the whole modeline.  Unfortunately replacing the modeline hides useful information when creating screencasts and code demos with Spacemacs.  There seems to be less information provides too.

Further experiments with keycast show that using the tab-line approach would give a more appropriate result with the current implementation.  I'll submit a PR to add a key binding for the tab-line command for keycast to the command-log layer.

> I tried for a few hours to add keycast to the doom modeline but never had a good understanding of what I was doing.  I tried adding keycast after existing segments but I just kept getting errors saying they didnt exist.  I am missing some fundamental understanding of how the modeline works.  My original suggestion in this thread no longer works either (although I did get it to work in Doom Emacs correctly).

As keycast in tabline mode meets my needs then I think this issue can be closed.  In fact having keycast in the tabline is simpler to track as it says in one place.  The modeline moves around when which-key shows and mini-buffer is expanded, making keycast output harder to follow

Modeline is supported although keycast replaces the whole modeline with itself, so some enhancement would be merited if using keycast in the modeline is desired.
